File: javascript/sandbox-setup.js
#!/usr/bin/env node
// sandbox-setup.sh — portiert nach javascript
// Quelle: shell, Onboarding@main:scripts/sandbox-setup.sh
// Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

// Provisioniert die Claude-Code-Sandbox (Remote-Umgebung) reproduzierbar:
//   - Node-Dependencies (Frontend, npm)
//   - Python-Dependencies (Backend inkl. pytest)
//   - Medien-Tools: ffmpeg, ImageMagick, GIMP, Blender headless (apt) —
//     Fehlschlag blockiert die Session nicht; --skip-heavy laesst GIMP/Blender aus
// Idempotent: bereits Vorhandenes wird uebersprungen; der Container-Cache der
// Umgebung macht die apt-Installation zum Einmal-Aufwand.

const fs = require('fs');
const path = require('path');
const { spawnSync, execSync } = require('child_process');

const SKIP_HEAVY = process.argv.includes('--skip-heavy') ? 1 : 0;

process.chdir(path.join(__dirname, '..'));

function log(message) {
  console.log(`[sandbox-setup] ${message}`);
}

function runCommand(command, args = [], options = {}) {
  const result = spawnSync(command, args, { encoding: 'utf8', ...options });
  return {
    success: result.status === 0,
    output: result.stdout || result.stderr,
    error: result.error
  };
}

function commandExists(command) {
  return runCommand('which', [command]).success;
}

function getVersion(command, versionArg = '--version') {
  try {
    const result = spawnSync(command, [versionArg], { encoding: 'utf8' });
    return result.stdout.split('\n')[0] || result.stderr.split('\n')[0] || '';
  } catch (error) {
    return '';
  }
}

log("Node-Dependencies (npm install) …");
let result = runCommand('npm', ['install', '--no-audit', '--no-fund']);
if (!result.success) {
  log("FEHLER: npm install fehlgeschlagen");
  process.exit(1);
}

log("Python-Dependencies (backend/requirements-dev.txt) …");
result = runCommand('pip3', ['install', '--quiet', '-r', 'backend/requirements-dev.txt']);
if (!result.success) {
  log("FEHLER: pip install fehlgeschlagen");
  process.exit(1);
}

let APT_UPDATED = false;

function aptInstall(pkg, bin) {
  if (commandExists(bin)) {
    const versionInfo = getVersion(bin, ['-version']) || getVersion(bin, ['--version']) || '';
    log(`${pkg} bereits vorhanden (${versionInfo})`);
    return true;
  }
  
  log(`Installiere ${pkg} …`);
  try {
    if (!APT_UPDATED) {
      execSync('DEBIAN_FRONTEND=noninteractive apt-get update -qq', { stdio: 'pipe' });
      APT_UPDATED = true;
    }
    execSync(`DEBIAN_FRONTEND=noninteractive apt-get install -y -qq ${pkg}`, { stdio: 'pipe' });
    return true;
  } catch (error) {
    log(`WARNUNG: ${pkg} konnte nicht installiert werden (Netzwerk-Policy?) — Medien-Schritte ggf. eingeschraenkt`);
    return false;
  }
}

aptInstall('ffmpeg', 'ffmpeg');
aptInstall('imagemagick', 'convert');
if (SKIP_HEAVY === 0) {
  aptInstall('gimp', 'gimp');
  aptInstall('blender', 'blender');
}

// Visual QA tools
aptInstall('xvfb', 'Xvfb');
aptInstall('x11-utils', 'xdpyinfo');
aptInstall('libnss3-tools', 'certutil');

if (!commandExists('google-chrome-stable')) {
  log("Installiere Google Chrome Stable …");
  try {
    const tmpDeb = '/tmp/chrome.deb';
    execSync(`curl -fsSL -o ${tmpDeb} https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb`, { stdio: 'pipe' });
    execSync(`DEBIAN_FRONTEND=noninteractive apt-get install -y -qq ${tmpDeb}`, { stdio: 'pipe' });
    const chromeVersion = getVersion('google-chrome-stable', ['--version']);
    if (chromeVersion) {
      log(`Chrome installiert: ${chromeVersion}`);
    } else {
      log("WARNUNG: Chrome-Installation fehlgeschlagen");
    }
    fs.unlinkSync(tmpDeb);
  } catch (error) {
    log("WARNUNG: Chrome-Download fehlgeschlagen (Netzwerk-Policy?)");
  }
}

// Proxy-CA in Chromes NSS-DB
if (commandExists('certutil') && fs.existsSync('/root/.ccr/ca-bundle.crt')) {
  const nssdbPath = path.join(process.env.HOME || '', '.pki/nssdb');
  try {
    fs.mkdirSync(nssdbPath, { recursive: true });
    execSync(`certutil -d sql:${nssdbPath} -N --empty-password`, { stdio: 'ignore' });
    
    try {
      execSync(`certutil -d sql:${nssdbPath} -L`, { stdio: 'pipe' });
      const certExists = execSync(`certutil -d sql:${nssdbPath} -L`, { encoding: 'utf8' }).includes('ccr-proxy-ca');
      if (!certExists) {
        execSync(`certutil -d sql:${nssdbPath} -A -t "C,," -n ccr-proxy-ca -i /root/.ccr/ca-bundle.crt`, { stdio: 'ignore' });
        log("Proxy-CA in Chrome-NSS-Store importiert");
      }
    } catch (error) {
      // continue regardless of error
    }
  } catch (error) {
    // continue regardless of error
  }
}

// Playwright installation
if (fs.existsSync('node_modules') && !fs.existsSync('node_modules/playwright')) {
  if (commandExists('npm')) {
    result = runCommand('npm', ['install', '--no-audit', '--no-fund', '--no-save', 'playwright'], { stdio: 'pipe' });
    if (result.success) {
      log("Playwright (Node) installiert");
    } else {
      log("WARNUNG: Playwright-npm-Install fehlgeschlagen");
    }
  }
}

// Git configuration
try {
  if (spawnSync('git', ['rev-parse', '--is-inside-work-tree'], { stdio: 'ignore' }).status === 0) {
    const pwd = process.cwd();
    execSync(`git config credential."https://x-access-token@github.com".helper "!${pwd}/.claude/git-credential-pat.sh"`, { stdio: 'ignore' });
    execSync('git remote set-url --push origin "https://x-access-token@github.com/KikiKari/Onboarding.git"', { stdio: 'ignore' });
    log("Git-Push-Route: direkt zu github.com (PAT via Credential-Helper)");
  }
} catch (error) {
  // continue regardless of error
}

// Docker daemon
if (commandExists('dockerd') && spawnSync('docker', ['info'], { stdio: 'ignore' }).status !== 0) {
  log("Starte Docker-Daemon (Registry-Mirror: mirror.gcr.io) …");
  try {
    fs.mkdirSync('/etc/docker', { recursive: true });
    const daemonJsonPath = '/etc/docker/daemon.json';
    if (!fs.existsSync(daemonJsonPath)) {
      fs.writeFileSync(daemonJsonPath, '{"registry-mirrors":["https://mirror.gcr.io"]}');
    }
    
    // Start docker daemon in background
    spawnSync('dockerd', [], { stdio: 'ignore', detached: true });
    
    // Wait for docker to start
    let dockerStarted = false;
    for (let i = 0; i < 15; i++) {
      if (spawnSync('docker', ['info'], { stdio: 'ignore' }).status === 0) {
        dockerStarted = true;
        break;
      }
      execSync('sleep 1', { stdio: 'ignore' });
    }
    
    if (dockerStarted) {
      log("Docker-Daemon laeuft");
    } else {
      log("WARNUNG: Docker-Daemon nicht gestartet");
    }
  } catch (error) {
    log("WARNUNG: Docker-Daemon nicht gestartet");
  }
}

log("Fertig. Versionen:");
console.log(`[sandbox-setup]   node ${getVersion('node')}`);
console.log(`[sandbox-setup]   ${getVersion('python3')}`);

if (commandExists('ffmpeg')) {
  console.log(`[sandbox-setup]   ${getVersion('ffmpeg', ['-version']).split('\n')[0]}`);
}
if (commandExists('convert')) {
  console.log(`[sandbox-setup]   ${getVersion('convert', ['-version']).split('\n')[0]}`);
}
if (commandExists('gimp')) {
  console.log(`[sandbox-setup]   ${getVersion('gimp', ['--version']).split('\n')[0]}`);
}
if (commandExists('blender')) {
  console.log(`[sandbox-setup]   ${getVersion('blender', ['--version']).split('\n')[0]}`);
}

process.exit(0);

File: python/sandbox-setup.py
#!/usr/bin/env python3
# sandbox-setup.sh — portiert nach python
# Quelle: shell, Onboarding@main:scripts/sandbox-setup.sh
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

import os
import sys
import subprocess
import shutil
import tempfile
import time

def log(message):
    print(f'[sandbox-setup] {message}')

def run_command(cmd, check=True, capture_output=False):
    try:
        if capture_output:
            result = subprocess.run(cmd, shell=True, check=check, capture_output=True, text=True)
            return result.stdout.strip()
        else:
            subprocess.run(cmd, shell=True, check=check)
            return None
    except subprocess.CalledProcessError as e:
        if check:
            raise
        return None

def command_exists(command):
    return shutil.which(command) is not None

def get_command_version(command, version_flag='--version'):
    try:
        result = subprocess.run([command, version_flag], capture_output=True, text=True)
        return result.stdout.split('\n')[0] if result.stdout else 'unknown version'
    except:
        return 'unknown version'

def apt_install(pkg, binary_name):
    if command_exists(binary_name):
        version_info = get_command_version(binary_name)
        log(f"{pkg} bereits vorhanden ({version_info})")
        return True
    
    log(f"Installiere {pkg} …")
    
    # Update package list if not already done
    if not hasattr(apt_install, 'apt_updated'):
        apt_install.apt_updated = False
    
    if not apt_install.apt_updated:
        env = os.environ.copy()
        env['DEBIAN_FRONTEND'] = 'noninteractive'
        try:
            subprocess.run(['apt-get', 'update', '-qq'], env=env, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            apt_install.apt_updated = True
        except subprocess.CalledProcessError:
            log("WARNUNG: apt-get update fehlgeschlagen")
            return False
    
    # Install package
    env = os.environ.copy()
    env['DEBIAN_FRONTEND'] = 'noninteractive'
    try:
        subprocess.run(['apt-get', 'install', '-y', '-qq', pkg], env=env, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        return True
    except subprocess.CalledProcessError:
        log(f"WARNUNG: {pkg} konnte nicht installiert werden (Netzwerk-Policy?) — Medien-Schritte ggf. eingeschraenkt")
        return False

def main():
    skip_heavy = len(sys.argv) > 1 and sys.argv[1] == "--skip-heavy"
    
    # Change to parent directory of this script
    script_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(os.path.join(script_dir, ".."))
    
    # Node dependencies
    log("Node-Dependencies (npm install) …")
    try:
        run_command("npm install --no-audit --no-fund")
    except subprocess.CalledProcessError:
        log("FEHLER: npm install fehlgeschlagen")
        sys.exit(1)
    
    # Python dependencies
    log("Python-Dependencies (backend/requirements-dev.txt) …")
    try:
        run_command("pip3 install --quiet -r backend/requirements-dev.txt")
    except subprocess.CalledProcessError:
        log("FEHLER: pip install fehlgeschlagen")
        sys.exit(1)
    
    # Media tools
    apt_install('ffmpeg', 'ffmpeg')
    apt_install('imagemagick', 'convert')
    
    if not skip_heavy:
        apt_install('gimp', 'gimp')
        apt_install('blender', 'blender')
    
    # Visual QA tools
    apt_install('xvfb', 'Xvfb')
    apt_install('x11-utils', 'xdpyinfo')
    apt_install('libnss3-tools', 'certutil')
    
    if not command_exists('google-chrome-stable'):
        log("Installiere Google Chrome Stable …")
        with tempfile.NamedTemporaryFile(suffix='.deb', delete=False) as tmp_deb:
            tmp_deb_path = tmp_deb.name
        
        try:
            subprocess.run(['curl', '-fsSL', '-o', tmp_deb_path, 'https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb'], check=True)
            env = os.environ.copy()
            env['DEBIAN_FRONTEND'] = 'noninteractive'
            subprocess.run(['apt-get', 'install', '-y', '-qq', tmp_deb_path], env=env, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            chrome_version = run_command("google-chrome-stable --version", capture_output=True)
            log(f"Chrome installiert: {chrome_version}")
        except subprocess.CalledProcessError:
            log("WARNUNG: Chrome-Installation fehlgeschlagen")
        finally:
            if os.path.exists(tmp_deb_path):
                os.unlink(tmp_deb_path)
    
    # Proxy CA import for Chrome
    if command_exists('certutil') and os.path.exists('/root/.ccr/ca-bundle.crt'):
        nssdb_path = os.path.expanduser('~/.pki/nssdb')
        os.makedirs(nssdb_path, exist_ok=True)
        
        try:
            subprocess.run(['certutil', '-d', f'sql:{nssdb_path}', '-N', '--empty-password'], 
                          stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        except:
            pass  # Ignore errors
        
        try:
            result = subprocess.run(['certutil', '-d', f'sql:{nssdb_path}', '-L'], 
                                  capture_output=True, text=True)
            if 'ccr-proxy-ca' not in result.stdout:
                subprocess.run(['certutil', '-d', f'sql:{nssdb_path}', '-A', '-t', 'C,,', '-n', 'ccr-proxy-ca', '-i', '/root/.ccr/ca-bundle.crt'],
                              stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
                log("Proxy-CA in Chrome-NSS-Store importiert")
        except:
            pass  # Ignore errors
    
    # Playwright installation
    if os.path.exists('node_modules') and not os.path.exists('node_modules/playwright'):
        if command_exists('npm'):
            try:
                subprocess.run(['npm', 'install', '--no-audit', '--no-fund', '--no-save', 'playwright'],
                             stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
                log("Playwright (Node) installiert")
            except subprocess.CalledProcessError:
                log("WARNUNG: Playwright-npm-Install fehlgeschlagen")
    
    # Git push configuration
    try:
        if subprocess.run(['git', 'rev-parse', '--is-inside-work-tree'], 
                          capture_output=True, check=True).returncode == 0:
            current_dir = os.getcwd()
            credential_helper = f"!{current_dir}/.claude/git-credential-pat.sh"
            
            subprocess.run(['git', 'config', f'credential.https://x-access-token@github.com.helper', credential_helper])
            subprocess.run(['git', 'remote', 'set-url', '--push', 'origin', 'https://x-access-token@github.com/KikiKari/Onboarding.git'])
            log("Git-Push-Route: direkt zu github.com (PAT via Credential-Helper)")
    except:
        pass  # Ignore git configuration errors
    
    # Docker daemon setup
    if command_exists('dockerd') and not command_exists('docker') or not run_command("docker info", check=False, capture_output=True):
        log("Starte Docker-Daemon (Registry-Mirror: mirror.gcr.io) …")
        os.makedirs('/etc/docker', exist_ok=True)
        
        daemon_config = '/etc/docker/daemon.json'
        if not os.path.exists(daemon_config):
            with open(daemon_config, 'w') as f:
                f.write('{"registry-mirrors":["https://mirror.gcr.io"]}')
        
        # Start docker daemon in background
        subprocess.Popen(['dockerd'], stdout=open('/tmp/dockerd.log', 'w'), stderr=subprocess.STDOUT)
        
        # Wait for docker to start
        for _ in range(15):
            try:
                subprocess.run(['docker', 'info'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
                break
            except subprocess.CalledProcessError:
                time.sleep(1)
        
        try:
            subprocess.run(['docker', 'info'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
            log("Docker-Daemon laeuft")
        except subprocess.CalledProcessError:
            log("WARNUNG: Docker-Daemon nicht gestartet")
    
    # Print versions summary
    log("Fertig. Versionen:")
    try:
        node_version = run_command("node --version", capture_output=True)
        log(f"  node {node_version}")
    except:
        pass
    
    try:
        python_version = run_command("python3 --version", capture_output=True)
        log(f"  {python_version}")
    except:
        pass
    
    if command_exists('ffmpeg'):
        try:
            ffmpeg_version = subprocess.run(['ffmpeg', '-version'], capture_output=True, text=True)
            log(f"  {ffmpeg_version.stdout.split(chr(10))[0]}")
        except:
            pass
    
    if command_exists('convert'):
        try:
            convert_version = subprocess.run(['convert', '-version'], capture_output=True, text=True)
            log(f"  {convert_version.stdout.split(chr(10))[0]}")
        except:
            pass
    
    if command_exists('gimp'):
        try:
            gimp_version = subprocess.run(['gimp', '--version'], capture_output=True, text=True)
            log(f"  {gimp_version.stdout.split(chr(10))[0]}")
        except:
            pass
    
    if command_exists('blender'):
        try:
            blender_version = subprocess.run(['blender', '--version'], capture_output=True, text=True)
            log(f"  {blender_version.stdout.split(chr(10))[0]}")
        except:
            pass

if __name__ == "__main__":
    main()
